Classes/Controller/TableOfContentsController.php

```diff
@@ -34,6 +34,156 @@ class TableOfContentsController extends AbstractController
+            if (
+                !empty($this->requestData['page'])
+                && !empty($this->document->getDoc()->physicalStructure)
+            ) {
+                $this->activeEntries = array_merge((array) $this->document->getDoc()->smLinks['p2l'][$this->document->getDoc()->physicalStructure[0]],
+                    (array) $this->document->getDoc()->smLinks['p2l'][$this->document->getDoc()->physicalStructure[$this->requestData['page']]]);
+                if (
+                    !empty($this->requestData['double'])
+                    && $this->requestData['page'] < $this->document->getDoc()->numPages
+                ) {
+                    $this->activeEntries = array_merge($this->activeEntries,
+                        (array) $this->document->getDoc()->smLinks['p2l'][$this->document->getDoc()->physicalStructure[$this->requestData['page'] + 1]]);
+                }
+            }
+            // Go through table of contents and create all menu entries.
+            foreach ($this->document->getDoc()->tableOfContents as $entry) {
+                $menuArray[] = $this->getMenuEntry($entry, true);
+            }
+        } else {
+            // Go through table of contents and create top-level menu entries.
+            foreach ($this->document->getDoc()->tableOfContents as $entry) {
+                $menuArray[] = $this->getMenuEntry($entry, false);
+            }
+            // Build table of contents from database.
+            $result = $this->documentRepository->getTableOfContentsFromDb($this->document->getUid(), $this->document->getPid(), $this->settings);
+
+            $allResults = $result->fetchAll();
+
+            if (count($allResults) > 0) {
+                $menuArray[0]['ITEM_STATE'] = 'CURIFSUB';
+                $menuArray[0]['_SUB_MENU'] = [];
+                foreach ($allResults as $resArray) {
+                    $entry = [
+                        'label' => !empty($resArray['mets_label']) ? $resArray['mets_label'] : $resArray['title'],
+                        'type' => $resArray['type'],
+                        'volume' => $resArray['volume'],
+                        'orderlabel' => $resArray['mets_orderlabel'],
+                        'pagination' => '',
+                        'targetUid' => $resArray['uid']
+                    ];
+                    $menuArray[0]['_SUB_MENU'][] = $this->getMenuEntry($entry, false);
+                }
+            }
+        }
+        return $menuArray;
+    }
+
+    /**
+     * This builds a menu for list of 3D objects
+     *
+     * @access protected
+     *
+     * @param string $content: The PlugIn content
+     * @param array $conf: The PlugIn configuration
+     *
+     * @return array HMENU array
+     */
+    protected function makeMenuFor3DObjects()
+    {
+        $menuArray = [];
+        // Is the document an external file?
+        if (!MathUtility::canBeInterpretedAsInteger($this->requestData['id'])) {
```

Comment:
Why do you check for external files here? Isn't this supposed to work for indexed files as well?

Reply:
Currently it works only with external files. I don't think that currently indexed files would contain all needed information to display.

Reply:
What's the difference here between an external file and an indexed file? Since the METS files don't get changed when indexing, this should make no difference. Or am I missing something?

Reply:
I'm not sure. Te whole METS gets indexed or only defined fields?

Reply:
Only the configured fields get extracted from the METS and inserted into the Solr index and/or database. Thus, only those fields can be searched or displayed in the list view.

But as soon as you are working within a single document (i. e. page view, metadata view, table of contents, etc.) the original METS file is loaded anyway - regardless if the document is indexed or external.